pippo-core/src/main/java/ro/fortsoft/pippo/core/Response.java

```diff
@@ -388,14 +388,64 @@ public boolean isCommitted() {
         return httpServletResponse.isCommitted();
     }
 
+    /**
+     * Attempts to negotiate the Response content type based on the Request
+     * accept type. If an agreement can't be reached, the Response content-type
+     * is not set.
+     *
+     * This allows a response chain such as
+     * <code>response.xml().negotiate(request).send(myObject);</code> which will
+     * set the content type as <code>application/xml</code> and optionally
+     * override it after negotation.
+     *
+     * @param request
+     * @return the response
+     */
+    public Response negotiate(Request request) {
```

Comment:
Yes, it's exactly for REST API and after thinking about is some more I may have to adjust my implementation and revise the fix I implemented in a97cb2a.

Let's say I want to serve objects for both reads and writes to an API.  I have a few couple choices here.

1. response.send(contact, fixedContentType).
2. response.send(contact, request.getAcceptType())
3. response.send(result, request.getContentType())

The first one is easy, Pippo dictates content type for all responses.  That's fine if the API serves only one specific type (i.e. JSON).

In the second one, Pippo sends the object using the type preferred by the `Request` based on the `Accept` header.  This is the GET/OPTIONS/HEAD scenario.  If the `Accept` header is malformed, empty, or does not report a known content type, we throw a server-side exception which results in an internal error (500) being sent to the client.

In the third one, Pippo sends the object using the type preferred by the `Request` based on the `Content-Type` header of the request.  This is the PUT/POST/PATCH scenario.

The `negotiate` method should be used to gracefully figure out which type to use when sending the object back to the client.  This method needs to consider the HTTP method, first, to determine the header check order.  If I PUT as JSON, then I should receive JSON back.  That's the mistake I made in a97cb2a.  Scenario 2 was failing so I reversed the code but in doing that I think I broke scenario 3.

If you search for *content negotiation* with your favorite search engine you find lots of resources.

So, yes the final value of contentType is not 100% predictable.  But it is guaranteed to be either your specified default (`xml()`) or one of your registered content type engines via `negotiate()`.